featurebyte/query_graph/transform/sdk_code.py

```diff
@@ -40,11 +39,137 @@ class SDKCodeGlobalState(BaseModel):
+            # `<CONDITIONAL> -> <ASSIGN> -> <ASSIGN>` but not `<CONDITIONAL> -> <ASSIGN>`.
+            #
+            # On the other hand, if `view` doesn't have `<col_name>` column, then `view[<col_name>][<mask>]`
+            # will throw an error. In this case, we should use the second case to generate SDK code.
+            assign_node = forward_nodes[0]
+            assign_node_input_node_names = query_graph.backward_edges_map[assign_node.name]
+            assign_input_view_op_struct = node_name_to_operation_structure[
+                assign_node_input_node_names[0]
+            ]
+            assign_input_view_columns = assign_input_view_op_struct.output_column_names
+            assign_column_name = assign_node.parameters.name  # type: ignore
+            if assign_column_name in assign_input_view_columns:
+                # This handle the case `view[<col>][<mask>] = <value>`. Since this SDK code actually contains
+                # ConditionalNode and AssignNode, we need to output info dict for the ConditionalNode and
+                # postpone the assignment operation to AssignNode.
+                self.as_info_dict_node_names.add(node.name)
+
+    def _identify_no_op_node(self, node: Node, backward_nodes: List[Node]) -> None:
+        """
+        Identify a node as no-op node by looking at the structure of the node type & its input node types.
+
+        Parameters
+        ----------
+        node: Node
+            Node to check
+        backward_nodes: List[Node]
+            Backward nodes of the node
+        """
+        if node.type == NodeType.PROJECT and backward_nodes[0].type in (
+            NodeType.ITEM_GROUPBY,
+            NodeType.AGGREGATE_AS_AT,
+        ):
+            # ItemGroupByNode's & AggregateAsAtNode's SDK code like
+            # * `view.groupby(...).aggregate(...)`
+            # * `view.groupby(...).aggregate_as_at(...)`
+            # already return a series, there is no operation for project node to generate any SDK code.
+            self.no_op_node_names.add(node.name)
+
+    def _identify_required_copy_node(
+        self, node: Node, backward_nodes: List[Node], edges_map: Dict[str, List[str]]
+    ) -> None:
+        """
+        Identify whether a copy is required for the inplace operation node.
+
+        Parameters
+        ----------
+        node: Node
+            Node to check
+        backward_nodes: List[Node]
+            Backward nodes of the node
+        edges_map: Dict[str, str]
+            Edges map of the query graph
+        """
+        if (
+            node.is_inplace_operation_in_sdk_code
+            and len(set(edges_map[backward_nodes[0].name])) > 1
+        ):
+            # If there are more than one distinct output nodes from the first input, it means that the backward node
+            # is used by other nodes. In this case, we need to copy the backward node to avoid inplace operation
```

Comment:
By `copy the backward node`, does it actually mean copying the node or the view? I think it's better to update the comment since it gets a bit confusing.

Reply:
It should be the input `view` but not `backward node`. Will update this part.